drift_dev/lib/src/writer/tables/data_class_writer.dart

```diff
@@ -269,8 +269,9 @@ class DataClassWriter {
         // apply type converter before writing the variable
         final converter = column.typeConverter;
         final fieldName = converter!.tableAndField;
-        final assertNotNull = !column.nullable && scope.generationOptions.nnbd;
-
+        final assertNotNull = !column.nullable &&
```

Comment:
Same here, maybe we don't need this check anymore.

Reply:
> This prevent user to create a single TypeConverter for nullable and non-nullable versions of the same type.

My idea is that drift would do that for users. They just write the non-nullable type converter. If this converter is applied to a nullable column, drift wraps that in a nullable converter behind the scenes.

> We assume user know what he do.

I think I understand the reasoning, but I don't fully agree with it. One of the reasons null safety is a popular Dart feature is that, without deep integration into the type system, dealing with nulls is really hard and error-prone. If we add non-null assertions everywhere and effectively trust the user to know what they're doing, aren't we thwarting null-safety with that?

I'll take a guess and assume that almost all type converters are written in a way that they are declared with non-nullable types. And yet, users currently have to guard for `null` (or use a `NullAwareTypeConverter`) even though their converters were never meant to map `null` values in the first place. I think the current behavior fixes that:

- if you apply a non-nullable type converter to a non-nullable column, it will always be used.
- if you apply a non-nullable type converter to a nullable column, it will be skipped for `null`s (which aren't mapped); it will be used for other values.

With some changes I made in bfad77788e4451633a8999271a48d613991b5493, this now works for compiled queries too (where a non-nullable column could be nullable because it's coming from an `OUTER JOIN` for instance). In my opinion, getting this case right is the most important thing.

Reply:
That was a good idea :) Perfect! Thanks